src/peer_connection.cpp

```diff
@@ -355,6 +355,18 @@ namespace libtorrent {
 					peer_log(peer_log_alert::outgoing, "SET_TOS", "tos: %d e: %s"
 						, m_settings.get_int(settings_pack::peer_tos), ec.message().c_str());
 				}
+#endif
+#ifdef IP_DSCP_TRAFFIC_TYPE
+				ec.clear();
+				m_socket->set_option(dscp_traffic_type(char(m_settings.get_int(settings_pack::peer_tos) >> 2)), ec);
```

Comment:
note the `>> 2` here

Reply:
Anal nit - you're quietly clamping the other socket options to 8 bits. To match, this one should be quietly clamped to 6 bits, so `dscp_traffic_type(char(m_settings.get_int(settings_pack::peer_tos)) >> 2)`. As it is, you're passing bits 2-9 of the settings value.

Reply:
I'm uncomfortable to right-shift a signed type, but I can truncate it with `&`

Reply:
Good catch. Yes, I was assuming char was unsigned.

How's about clamping the TOS values to 6 bits with `&` too?

`traffic_class(m_settings.get_int(settings_pack::peer_tos) & 0xfc)`

Save the network from people getting confused and putting a DSCP code into peer_tos.

Reply:
Actually, the `char` is wrong even as it is. If `char` is signed, that will turn `0xc0` into int `-0x40` (`0xFFFFFFC0`), and that will get passed as the socket option. Kernel may well reject that. Certainly `-1` has a special meaning for `IPV6_TCLASS`, so it won't just be looking at the bottom 8 bits.

Probably haven't hit that because the top-bit-set DSCPs are higher priority and probably not going to be selected for torrents.